File: questions/q134_maximum_stream_subarray/heaps.py
import heapq

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        store = []
        ans = []

        for i in range(k) :
            heapq.heappush(store, (-1*nums[i], i))
        ans.append(-1 * store[0][0])

        for i in range(k, len(nums)) :
            heapq.heappush(store, (-1*nums[i], i))
            while store[0][1] <= i-k :
                heapq.heappop(store)
            
            ans.append(-1 * store[0][0])
        
        return ans

File: questions/q134_maximum_stream_subarray/heaps.cpp
// This is for the leetcode version of the problem


class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        priority_queue<pair<int, int>, vector<pair<int, int>>> store;
        vector<int> ans;

        for (int i=0; i<k; i++) {
            store.push({nums[i], i});
        }
        ans.push_back(store.top().first);

        for (int i=k; i<nums.size(); i++) {
            store.push({nums[i], i});

            while (store.top().second <= i-k) {
                store.pop();
            }
            ans.push_back(store.top().first);
        }

        return ans;
    }
};
